submission/races.go

```diff
@@ -0,0 +1,299 @@
+// the entry wasn't requested (added) before.
+func (sub *safeSubmissionState) request(logURL string, cancel context.CancelFunc) bool {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	if _, requested := sub.results[logURL]; requested {
+		// Already requested.
+		return false
+	}
+	sub.results[logURL] = &submissionResult{}
+	isAwaitedByGroup := false
+	for g := range sub.logToGroups[logURL] {
+		if sub.groupNeeds[g] > 0 {
+			isAwaitedByGroup = true
+			break
+		}
+	}
+	if !isAwaitedByGroup {
+		// No groups expecting result from this Log.
+		return false
+	}
+	sub.cancels[logURL] = cancel
+	return true
+}
+
+// returns whether the Log has already been requested.
+func (sub *safeSubmissionState) has(logURL string) bool {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	_, requested := sub.results[logURL]
+	return requested
+}
+
+// Handles SCT-result. Writes it down if it is an error or awaited-SCT.
+// Re-calculates group-completion and cancels any running but not-awaited-anymore
+/// Log-requests.
+func (sub *safeSubmissionState) handleResult(logURL string, sct *ct.SignedCertificateTimestamp, err error) {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	if sct == nil {
+		sub.results[logURL] = &submissionResult{sct: sct, err: err}
+		return
+	}
+	// If at least one group needs that SCT, result is set. Otherwise ignored.
+	for groupName := range sub.logToGroups[logURL] {
+		if sub.groupNeeds[groupName] > 0 {
+			sub.results[logURL] = &submissionResult{sct: sct, err: err}
+		}
+		sub.groupNeeds[groupName]--
+	}
+
+	// Cancel any pending Log-requests for which there's no more awaiting Log-group.
+	for logURL, groupSet := range sub.logToGroups {
+		isAwaitedByGroup := false
+		for g := range groupSet {
+			if sub.groupNeeds[g] > 0 {
+				isAwaitedByGroup = true
+				break
+			}
+		}
+		if !isAwaitedByGroup && sub.cancels[logURL] != nil {
+			sub.cancels[logURL]()
+			delete(sub.cancels, logURL)
+		}
+	}
+}
+
+// returns whether the group has stopped waiting for more SCTs.
+func (sub *safeSubmissionState) groupComplete(groupName string) bool {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	needs, ok := sub.groupNeeds[groupName]
+	if !ok {
+		return true
+	}
+	return needs <= 0
+}
+
+func (sub *safeSubmissionState) collectSCTs() []*AssignedSCT {
+	sub.mu.Lock()
+	defer sub.mu.Unlock()
+	scts := []*AssignedSCT{}
+	for logURL, r := range sub.results {
+		if r != nil && r.sct != nil {
+			scts = append(scts, &AssignedSCT{LogURL: logURL, SCT: r.sct})
+		}
+	}
+	return scts
+}
+
+// submitInterval calculates duration for consequent call.
+// For first parallelStart calls duration is 0, while every next one gets
+// additional dur interval.
+func submitInterval(idx int, parallelStart int, dur time.Duration) time.Duration {
+	if idx < parallelStart {
+		return time.Duration(0)
+	}
+	return time.Duration(idx+1-parallelStart) * dur
+}
+
+// Submitter is interface wrapping Log-request-response cycle and any processing.
```

Comment:
Please be more specific, e.g.:
`// Submitter allows sending certificate chains to CT logs.`

Also, I think it makes sense to put this interface to the top, as it's essentially one of the entry points to this package.

Reply:
Are you sure your description is more specific?
There's no Go-recs on func location.

Reply:
Yes, I am sure. CT logs have [8 endpoints](https://tools.ietf.org/html/rfc6962#section-4). This interface covers only `add-[pre-]chain`, which I suggest to emphasize in the comment. "Log-request-responce cycle" is not quite specific because any RPC is a request-responce cycle.